### ISSUE TITLE
[Proposal] When you change ember versions try to stay on the same section of the guide

### ISSUE THREAD

**Issue by varblob** [emberjs/guides#206]
Currently when you change the version of the ember guide it resets you to the root of the guide, which is not ideal.  

I have a couple proposals for how to try and approach this with the full expectation that my understanding of how things are setup maybe totally wrong.

# The simple fallback approach

When a user switches version attempt to match based on the title of the guide, if this fails match based on the section, if this fails goto root.  

It may also make sense to add some sort of non title identifier to guide pages so small title changes don't cause fallback.

# The manual mapping approach

In the data.yml if a guide goes away or changes in some significant way then put a manual mapping of the closest page.  I would imagine that there should always be some sort of mapping possible since during the deprecation of the old feature there is normally discussion of how to upgrade / update.  If there is no one to one mapping for a page then it would seem an excellent opportunity to put an upgrade guide page that explains to users the how to upgrade or simply points them to more than one guide if there is a one to many mapping.  

# Other side benefits

If the mapping issue is solved, especially if it's intermixed with pages helping with upgrading, a link to the new version can be placed more prominently on out of date pages encouraging users to upgrade.

**Comment by trek** [comment]
Hey @varblob we'd need to see concrete details on how something like this can be implemented to demonstrate the idea has been thought through at a technical level. It's a good idea, but needs _actionable_ steps for people help out or :+1: the work.

There are _many_ technical questions that would need identifying and answering before we could do this.

e.g.
>When a user switches version attempt to match based on the title of the guide

Match where? Where is this data stored and how? How is it looked up? 

> In the data.yml if a guide goes away or changes in some significant way then put a manual mapping of the closest page.

We remove occurs in `v1.11.0/foos`  and map to the closest page `v1.12.0/new-foos`, which seems possibly easy? What happens when we've release five more revisions. Where is the data stored that the system knows  `v1.11.0/foos` now points to `v1.17.0/new-foos`. What if there is another change between the two? `v1.11.0/foos` -> `v1.13.0/new-foos` -> `v1.15.0/newer-and-imrproved-foos`. How does ` `v1.11.0/foos` know to point to `v1.15.0/newer-and-imrproved-foos`?

Where do we store data so it can mutate over time and remain manageable?

**Comment by varblob** [comment]
I totally understand the need for more detail I'm hoping this is the place to try and get there.

I will poke further at this in the evening.  Do you think you could point me towards a good entry point to try and understand better what is being done at the moment especially around how the mds are served up?

**Comment by trek** [comment]
> Do you think you could point me towards a good entry point to try and understand better what is being done at the moment especially around how the mds are served up between versions?

When we're ready to snapshot a version we run `middleman build` and move the resulting folder to https://github.com/emberjs/guides.emberjs.com as a "snapshot",  which is published to divshot.

**Comment by igorT** [comment]
Maybe the simplest approach that would cover most of the cases would be:
1. When generating the links to other versions, append the current guide url to the link:
    So if you are on `http://guides.emberjs.com/v1.10.0/routing/query-params/` the dropdown to 1.11.0 would link to `http://guides.emberjs.com/v1.11.0/routing/query-params/` 

2. Amend our 404 page to flash a message that the guide was not found and then redirect to the versioned root. Thus going to `http://guides.emberjs.com/v1.11.0/old-non-existent-guide` would redirect to  `http://guides.emberjs.com/v1.11.0` which would catch the problems when the guides move/rename/get deleted.

**Comment by trek** [comment]
>When generating the links to other versions, append the current guide url to the link:

Middleman/this repo is unaware of specific, but I think this approach is doable with some regexp trickery in https://github.com/emberjs/guides/blob/master/source/javascripts/app/guides/version-select.js.erb

**Comment by varblob** [comment]
I think I can take a whack at this but I'm not entirely sure how I should setup the divshot bit.  It seems I would need that to be able to test the version switching?  I cloned this repo but the urls don't include the version info.

**Comment by trek** [comment]
> I cloned this repo but the urls don't include the version info.

Correct. This project is not aware that the guides are versioned. It only knows about itself. That is the tricky bit.

**Comment by varblob** [comment]
Just a bump to say I would still be keen to work on this but could use some guidance on how to setup so I can test given divshot.  If it would be more useful to just make a go of it and submit the pr, I can do that to.

**Comment by michaelrkn** [comment]
@varblob there don't seem to be any automated tests currently for the JS. For manually testing, you could just make sure that it takes you to the URL you expect - eg, if you're at `http://localhost:4567/models/defining-models/` and then select 1.9 from the drop-down, the URL should become `http://localhost:4567/v1.9.0/models/defining-models/`. You'll see a 404 on the page, of course, but it will work correctly in production.

**Comment by nicolechung** [comment]
It looks like it's just jQuery in the erb file. Can't you just do something like:

```
versionSelect.on('change', function() {
      var version = this.value;
      var pathname = window.location.pathname;
      window.location = '/' + version + pathname;
    });
```

**Comment by michaelrkn** [comment]
Closed by #306. Can you please close @rwjblue?
